src/Microsoft.Health.Dicom.Functions/Indexing/ReindexOperation.cs

```diff
@@ -0,0 +1,141 @@
+        }
+
+        /// <summary>
+        ///  The orchestration function to add extended query tags.
+        /// </summary>
+        /// <param name="context">The context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns>The task.</returns>
+        [FunctionName(nameof(AddExtendedQueryTagsOrchestrationAsync))]
+        public async Task AddExtendedQueryTagsOrchestrationAsync(
+            [OrchestrationTrigger] IDurableOrchestrationContext context,
+            ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<AddExtendedQueryTagEntry>>();
+            var tagEntries = await context.CallActivityAsync<IEnumerable<ExtendedQueryTagStoreEntry>>(nameof(AddExtendedQueryTagAsync), input);
+            await context.CallSubOrchestratorAsync(nameof(ReindexExtendedQueryTagsOrchestrationAsync), context.InstanceId, tagEntries);
+        }
+
+        /// <summary>
+        /// The orchestration function to Reindex Extended Query Tags.
+        /// </summary>
+        /// <param name="context">the context.</param>
+        /// <param name="logger">The logger.</param>
+        /// <returns></returns>
+        [FunctionName(nameof(ReindexExtendedQueryTagsOrchestrationAsync))]
+        public async Task ReindexExtendedQueryTagsOrchestrationAsync(
+           [OrchestrationTrigger] IDurableOrchestrationContext context,
+           ILogger logger)
+        {
+            EnsureArg.IsNotNull(context, nameof(context));
+            logger = context.CreateReplaySafeLogger(logger);
+            var input = context.GetInput<IEnumerable<ExtendedQueryTagStoreEntry>>();
+
+
+            // TODO: should start a new orchestraion instead of while loop
+            while (true)
+            {
+                IReadOnlyList<long> watermarks = await context
+                    .CallActivityAsync<IReadOnlyList<long>>(nameof(GetNextWatermarksOfOperationAsync), context.InstanceId);
+                if (watermarks.Count == 0)
+                {
+                    break;
+                }
+
+                IReadOnlyList<ExtendedQueryTagStoreEntry> queryTags = await context
+                    .CallActivityAsync<IReadOnlyList<ExtendedQueryTagStoreEntry>>(nameof(GetExtendedQueryTagsOfOperationAsync), context.InstanceId);
+
+                if (queryTags.Count == 0)
+                {
+                    break;
+                }
+
+                // Reindex
+                // TODO: process them parallel
+                foreach (long watermark in watermarks)
+                {
+                    await context
+                        .CallActivityAsync(nameof(ReindexInstanceAsync), new ReindexInstanceInput() { TagEntries = queryTags, Watermark = watermark });
+                }
+
+                await context.CallActivityAsync(nameof(UpdateEndWatermarkOfOperationAsync),
```

Comment:
How does Update differ from Complete? Can they be the same?

Reply:
The Update update progress of operation, and underlying purely just update Endwatermark of operation. 
The Complete completes the operation. It deletes operation from TagOperationStore, and update Tag status to be `Ready` in TagStore.
They could be putting together, but I feel as for now it's more clear to separate them.

Reply:
I am unsure if we want the "complete" activity to delete the operation from the DB yet. We're already planning to have a job that regularly cleans up state from the history log. It seems reasonable it could perform this clean up too.

Reply:
we can talk about this tomorrow.